casper/src/test/resources/PoSTest.rho

```diff
@@ -3,32 +3,290 @@
+                                    ((initialUser1Balance - finalUser1Balance, "==", 100), "the user1 account decreases"),
+                                    ((initialUser2Balance - finalUser2Balance, "==", 200), "the user2 account decreases"),
+                                    ((finalPosBalance - initialPosBalance, "==", 300), "the pos account increases"),
+                                  ],
+                                  *ackCh
+                                )
+                              }
+                            }
+                          }
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          } |
+
+          contract test_pay_succeeds(rhoSpec, _, ackCh) = {
+            new setupCh, retCh, initialPosBalanceCh, finalPosBalanceCh,
+                initialUser1BalanceCh, finalUser1BalanceCh in {
+              prepareUser!("4444444444444444444444444444444444444444444444444444444444444444", 10000, *setupCh) |
+              for (@user1PubKey, _, @user1Vault <- setupCh) {
+                @posVault!("balance", *initialPosBalanceCh) |
+                @user1Vault!("balance", *initialUser1BalanceCh) |
+                for(@initialPosBalance <- initialPosBalanceCh;
+                    @initialUser1Balance <- initialUser1BalanceCh) {
+                  @PoS!("pay", 100, *retCh) |
+                  for ( @(true, _) <- retCh) {
+
+                    @posVault!("balance", *finalPosBalanceCh) |
+                    @user1Vault!("balance", *finalUser1BalanceCh) |
+                    for(@finalPosBalance <- finalPosBalanceCh;
+                        @finalUser1Balance <- finalUser1BalanceCh) {
+                      rhoSpec!("assertMany",
+                        [
+                          ((initialUser1Balance - finalUser1Balance, "==", 100), "the user account decreases"),
+                          ((finalPosBalance - initialPosBalance, "==", 100), "the pos account increases"),
+                        ],
+                        *ackCh
+                      )
+                    }
+                  }
+                }
+              }
+            }
+          } |
+
+          contract test_bonding_fails_if_deposit_fails(rhoSpec, _, ackCh) = {
+            new setupCh, retCh, bondsCh in {
+              prepareUser!("5555555555555555555555555555555555555555555555555555555555555555", 10000, *setupCh) |
+              for (_, _, _ <- setupCh) {
+                @PoS!("getBonds", *bondsCh) |
+                for (@initialBonds <- bondsCh) {
+                  @PoS!("bond", 20000, *retCh) |
+
+                  for ( @(result, msg) <- retCh) {
+                    @PoS!("getBonds", *bondsCh) |
+                    rhoSpec!("assertMany",
+                      [
+                        ((result, "==", false), "the bond should fail"),
```

Comment:
would be prudent to test the failure msg too

Reply:
I'm not happy with testing based on error messages but I agree that we should make sure the reason for the failure is the right one